Add media element watchtime diagnostic logging

jernoble opened the pull request:
#### 06391e123c0e3b40a30e1114af79cd61c4e2f603
<pre>
Add media element watchtime diagnostic logging
<a href="https://bugs.webkit.org/show_bug.cgi?id=277747">https://bugs.webkit.org/show_bug.cgi?id=277747</a>
<a href="https://rdar.apple.com/133396572">rdar://133396572</a>

Reviewed by Eric Carlson.

Add diagnostic logging of watchtime for various media properties. Specifically:
- Log watchtime separated by source type (HLS, live stream, media source, etc.)
- Log watchtime separated by video and audio codecs

For each of these watchtime log events, enable random sampling, so only a small
percentage of events will actually get logged. Additionally, break up watchtime
events into 5 minute intervals, which due to sampling, only some of which will
actually be logged.

* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::~HTMLMediaElement):
(WebCore::HTMLMediaElement::mediaPlayerRateChanged):
(WebCore::HTMLMediaElement::clearMediaPlayer):
(WebCore::HTMLMediaElement::stopWithoutDestroyingMediaPlayer):
(WebCore::HTMLMediaElement::startWatchtimeTimer):
(WebCore::HTMLMediaElement::pauseWatchtimeTimer):
(WebCore::HTMLMediaElement::invalidateWatchtimeTimer):
(WebCore::HTMLMediaElement::watchtimeTimerFired):
* Source/WebCore/html/HTMLMediaElement.h:
* Source/WebCore/page/DiagnosticLoggingKeys.cpp:
(WebCore::DiagnosticLoggingKeys::mediaSourceTypeWatchTimeKey):
(WebCore::DiagnosticLoggingKeys::mediaVideoCodecWatchTimeKey):
(WebCore::DiagnosticLoggingKeys::mediaAudioCodecWatchTimeKey):
(WebCore::DiagnosticLoggingKeys::secondsKey):
(WebCore::DiagnosticLoggingKeys::sourceTypeKey):
(WebCore::DiagnosticLoggingKeys::videoCodecKey):
(WebCore::DiagnosticLoggingKeys::audioCodecKey):
(WebCore::DiagnosticLoggingKeys::mediaElementSourceTypeDiagnosticLoggingKey):
* Source/WebCore/page/DiagnosticLoggingKeys.h:

Canonical link: <a href="https://commits.webkit.org/282004@main">https://commits.webkit.org/282004@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/945ce53485304b3951695238f68ae7259ff929f8

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/61629 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/40982 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/14220 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/65676 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/12174 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/63748 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/48667 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/12445 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/49766 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/8448 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/64698 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/38078 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/53381 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/30598 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/34738 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/10612 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/11105 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/56540 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/10912 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/67331 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/5572 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/10742 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/57102 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/5597 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/53331 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/57322 "Found 1 new API test failure: /WebKitGTK/TestUIClient:/webkit/WebKitWebView/file-chooser-request (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/13739 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/4584 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/36783 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/37867 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/38963 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/37612 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

eric-carlson commented on Source/WebCore/html/HTMLMediaElement.cpp:
```diff
@@ -9667,6 +9747,102 @@ void HTMLMediaElement::defaultSpatialTrackingLabelChanged(const String& defaultS
+        return;
+
+    // Silent, autoplaying content should not produce watchtime diagnostics:
+    if (!m_hasEverHadAudio || m_muted)
```
Should we call `watchtimeTimerFired()` when `muted` changes from false to true?

jernoble replied:
In the long run, I don't think it matters much, as we're trying to understand very broad, representative samples of watch time. If we miss out on a portion of the interval because a user muted the video half-way through, I think that's probably ok.

jernoble replied:
In fact, it will probably be balanced out by users who un-mute the video half-way through an interval!

jernoble committed:
Add media element watchtime diagnostic logging
https://bugs.webkit.org/show_bug.cgi?id=277747
rdar://133396572

Reviewed by Eric Carlson.

Add diagnostic logging of watchtime for various media properties. Specifically:
- Log watchtime separated by source type (HLS, live stream, media source, etc.)
- Log watchtime separated by video and audio codecs

For each of these watchtime log events, enable random sampling, so only a small
percentage of events will actually get logged. Additionally, break up watchtime
events into 5 minute intervals, which due to sampling, only some of which will
actually be logged.

* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::~HTMLMediaElement):
(WebCore::HTMLMediaElement::mediaPlayerRateChanged):
(WebCore::HTMLMediaElement::clearMediaPlayer):
(WebCore::HTMLMediaElement::stopWithoutDestroyingMediaPlayer):
(WebCore::HTMLMediaElement::startWatchtimeTimer):
(WebCore::HTMLMediaElement::pauseWatchtimeTimer):
(WebCore::HTMLMediaElement::invalidateWatchtimeTimer):
(WebCore::HTMLMediaElement::watchtimeTimerFired):
* Source/WebCore/html/HTMLMediaElement.h:
* Source/WebCore/page/DiagnosticLoggingKeys.cpp:
(WebCore::DiagnosticLoggingKeys::mediaSourceTypeWatchTimeKey):
(WebCore::DiagnosticLoggingKeys::mediaVideoCodecWatchTimeKey):
(WebCore::DiagnosticLoggingKeys::mediaAudioCodecWatchTimeKey):
(WebCore::DiagnosticLoggingKeys::secondsKey):
(WebCore::DiagnosticLoggingKeys::sourceTypeKey):
(WebCore::DiagnosticLoggingKeys::videoCodecKey):
(WebCore::DiagnosticLoggingKeys::audioCodecKey):
(WebCore::DiagnosticLoggingKeys::mediaElementSourceTypeDiagnosticLoggingKey):
* Source/WebCore/page/DiagnosticLoggingKeys.h:

Canonical link: https://commits.webkit.org/282004@main
